src/lib.rs

```diff
@@ -739,76 +597,260 @@ impl<'text> BidiInfo<'text> {
-    #[inline]
-    pub fn has_rtl(&self) -> bool {
-        level::has_rtl(&self.levels)
+    // This implementation is similar to the L2 implementation in `visual_runs()`
+    // but it cannot benefit from a precalculated LevelRun vector so needs to be different.
+
+    if levels.is_empty() {
+        return vec![];
+    }
+
+    // Get the min and max levels
+    let (mut min, mut max) = levels
+        .iter()
+        .fold((levels[0], levels[0]), |(min, max), &l| {
+            (cmp::min(min, l), cmp::max(max, l))
+        });
+
+    // Initialize an index map
+    let mut result: Vec<usize> = (0..levels.len()).collect();
+
+    if min == max && min.is_ltr() {
+        // Everything is LTR and at the same level, do nothing
+        return result;
+    }
+
+    // Stop at the lowest *odd* level, since everything below that
+    // is LTR and does not need further reordering
+    min = min.new_lowest_ge_rtl().expect("Level error");
+
+    // For each max level, take all contiguous chunks of
+    // levels ≥ max and reverse them
+    //
+    // We can do this check with the original levels instead of checking reorderings because all
+    // prior reorderings will have been for contiguous chunks of levels >> max, which will
+    // be a subset of these chunks anyway.
+    while min <= max {
+        let mut range = 0..0;
+        loop {
+            range = next_range(levels, range.end, max);
+            result[range.clone()].reverse();
+
+            if range.end >= levels.len() {
+                break;
+            }
+        }
+
+        max.lower(1).expect("Level error");
+    }
+
+    result
+}
+
+fn compute_bidi_info_for_para<D: BidiDataSource, T: TextSource + ?Sized>(
+    data_source: &D,
+    para: &ParagraphInfo,
+    is_pure_ltr: bool,
+    text: &T,
+    original_classes: &[BidiClass],
+    processing_classes: &mut [BidiClass],
+    levels: &mut Vec<Level>,
+) {
+    let new_len = levels.len() + para.range.len();
+    levels.resize(new_len, para.level);
+    if para.level == LTR_LEVEL && is_pure_ltr {
+        return;
+    }
+
+    let processing_classes = &mut processing_classes[para.range.clone()];
+    let levels = &mut levels[para.range.clone()];
+
+    explicit::compute(
+        text,
+        para.level,
+        original_classes,
+        levels,
+        processing_classes,
+    );
+
+    let sequences = prepare::isolating_run_sequences(para.level, original_classes, levels);
+    for sequence in &sequences {
+        implicit::resolve_weak(text, sequence, processing_classes);
+        implicit::resolve_neutral(
+            text,
+            data_source,
+            sequence,
+            levels,
+            original_classes,
+            processing_classes,
+        );
+    }
+    implicit::resolve_levels(processing_classes, levels);
+
+    assign_levels_to_removed_chars(para.level, original_classes, levels);
+}
+
+fn reorder_levels<T: TextSource + ?Sized>(
```

Comment:
nit: please also document these functions and what range of spec rules they follow